Adapt more forms to use `CdxFormHelper`

straight-shoota opened the pull request:
Applies the form field helper introduced in https://github.com/instedd/cdx/pull/1662 to more forms. This should finally resolve #1065.
<del>There are a couple of forms left which didn't seem to be trivial to adapt. We'll have to look into those more carefully.</del>
<ins>I changed it for all forms.</ins>

straight-shoota commented:
There are some issues with nested scopes (`#fields_for`). #1689 fixes that.

ysbaddaden commented on app/helpers/cdx_form_helper.rb:
```diff
@@ -78,30 +77,77 @@ def form_actions(&block)
+    nested_prefix = "#{nested_prefix}." if nested_prefix
     Hash[errors_to_show.map do |attribute|
-      messages = @object.errors.full_messages_for(attribute)
+      next if nested_prefix && !attribute.to_s.starts_with?(nested_prefix)
```
```suggestion
      next if nested_prefix && !attribute.to_s.start_with?(nested_prefix)
```

ysbaddaden commented on app/helpers/cdx_form_helper.rb:
```diff
@@ -78,30 +77,77 @@ def form_actions(&block)
+  def clear_errors_to_show
+    if nested_prefix = @options[:nested_errors_namespace]
+      nested_prefix = "#{nested_prefix}."
+      errors_to_show.delete_if { |key| key.to_s.starts_with?(nested_prefix) }
```
```suggestion
      errors_to_show.delete_if { |key| key.to_s.start_with?(nested_prefix) }
```

ysbaddaden commented on app/helpers/cdx_form_helper.rb:
```diff
@@ -78,30 +77,77 @@ def form_actions(&block)
+  def fields_for(record_name, record_object = nil, fields_options = {}, &block)
+    fields_options[:nested_errors_namespace] = prefixed_error_key(record_name)
+
+    super(record_name, record_object = nil, objectify_options(fields_options)) do |form|
```
```suggestion
    super(record_name, record_object, objectify_options(fields_options)) do |form|
```

ysbaddaden reviewed (commented):
Hum, I'm not so sure about `#fields_for` and nested keys. I need to check it up.

straight-shoota commented:
In light of https://github.com/instedd/cdx/pull/1689#pullrequestreview-989976210 I rebased this on main without #1689. It works with stock `fields_for` as well. Worst thing that happens is that errors from associated objects show up as form errors as well (besides directly on the field where they belong). But that's hardly a problem.
In the specs this happened only in a single form and there I added a simple fix to suppress the error for showing up on the parent form.

A proper fix should be implemented for `fields_for`, but we can already port the existing forms to use the form helper. It is clearly an improvement if error messages occasionally show up in two locations than all error messages being placed unrelated to the respective fields.

ysbaddaden commented on app/views/device_models/_form.haml:
```diff
@@ -1,90 +1,62 @@
-          - if !@device_model.new_record? && @device_model.setup_instructions.present?
+        - # FIXME: This is just to surpress the error message showing up in the
+        - # form (via `form_errors`).
+        - f.field_errors :"manifest.definition"
```
What about making this explicit with a nice helper?

```ruby
def FormFieldBuilder.skip_errors(attr_name)
  @object.errors.delete(attr_name.to_sym)
  nil
end
```

straight-shoota replied:
Yeah, that's a good idea. But such a skip method should probably delete all errors with that prefix path, not just exact matches. I wanted to keep that complexity out of this PR. Merging this is a great improvement and I wouldn't delay it. We can easily refactor it later (that's why I left the `FIXME`).

straight-shoota replied:
No, we really need to handle nested keys already for use case like this. We should skip all errors starting with the key of the nested form. I'm only explicitly skipping `:"manifest.definition"` for now because that's the only error that shows up in CI. If the manifest object contains other errors (and there are a couple of validations on that model), they would still show up in the parent, but we'd want to suppress them and only show them attached to the nested form.

ysbaddaden commented:
LGTM!
